### ISSUE TITLE
DataLoader with num_workers>0 fails when running with "spawn" and using joblib

### ISSUE THREAD

**Issue by gerardsn** [pytorch/pytorch#44687]
## 🐛 Bug

Running multiple jobs in parallel (using joblib) fails when num_workers in DataLoaders is > 0.

## To Reproduce

```python
import os, sys
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST
from torchvision import transforms
from joblib import Parallel, delayed

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

def main(num_workers, lr):
    data_set = MNIST(os.getcwd(), download=True, train=True, transform=transforms.ToTensor())
    data_loader = DataLoader(data_set, batch_size=64, num_workers=num_workers)
    model = Model().cuda()
    optim = torch.optim.Adam(model.parameters(), lr=lr)

    for batch in data_loader:
        x, y = batch
        y_hat = model(x.cuda())
        loss = F.cross_entropy(y_hat, y.cuda())
    print(f'completed {lr}')

if __name__ == '__main__':
    num_workers = int(sys.argv[1])
    lr_array = [1e-4, 1e-3, 1e-2]
    Parallel(n_jobs=3, backend='loky')(delayed(main)(num_workers, lr) for lr in lr_array)
```

This works:
```commandline
python code.py 0
```

This fails:
```commandline
python code.py 1
```

with error: (note that there are several times more cores available than required for this config.)

```commandline
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 253, in __call__
    for func, args, kwargs in self.items]
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 253, in <listcomp>
    for func, args, kwargs in self.items]
  File "MNIST_hydra.py", line 32, in main
    for batch in data_loader:
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 291, in __iter__
    return _MultiProcessingDataLoaderIter(self)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 737, in __init__
    w.start()
  File "/usr/lib/python3.7/multiprocessing/process.py", line 112, in start
    self._popen = self._Popen(self)
  File "/usr/lib/python3.7/multiprocessing/context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/backend/process.py", line 39, in _Popen
    return Popen(process_obj)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/backend/popen_loky_posix.py", line 52, in __init__
    self._launch(process_obj)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/backend/popen_loky_posix.py", line 157, in _launch
    pid = fork_exec(cmd_python, self._fds, env=process_obj.env)
AttributeError: 'Process' object has no attribute 'env'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "MNIST_hydra.py", line 41, in <module>
    Parallel(n_jobs=3, backend='loky')(delayed(main)(num_workers, lr) for lr in lr_array)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 1042, in __call__
    self.retrieve()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 921, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
AttributeError: 'Process' object has no attribute 'env'
(l2r) $ /opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/backend/resource_tracker.py:320: UserWarning: resource_tracker: There appear to be 33 leaked semlock objects to clean up at shutdown
  (len(rtype_registry), rtype))
```

## Expected behavior

Being able to set num_workers>0 while running multiple jobs in parallel with joblib.

## Environment

PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: TITAN Xp
Nvidia driver version: 440.100
cuDNN version: 7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] pytorch-lightning==0.9.1rc1
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] Could not collect

## Additional context

Same issue on GPU and CPU, and there are plenty of cores available for this to work.

cc @SsnL @VitalyFedyunin

**Comment by albanD** [comment]
Hi,

This errors seems to point inside joblib and not pytorch no?
Also maybe you need to change the default fork method?

**Comment by gerardsn** [comment]
@albanD Thanks for the suggestion. Setting the `DataLoader(..., multiprocessing_context='fork')` fixes the issue for me.  
However, when setting the fork method to `spawn` or `forkserver` I get the error below suggesting that there is a clash between `joblib`'s multiprocessing `backend='loky'` and the forking attempts by the DataLoader causing the workers to die. Not sure what the origin of this error is. Feel free to close this issue if the problem is not in the PyTorch domain.

```
Traceback (most recent call last):
  File "/usr/lib/python3.7/multiprocessing/forkserver.py", line 261, in main
    old_handlers)
  File "/usr/lib/python3.7/multiprocessing/forkserver.py", line 297, in _serve_one
    code = spawn._main(child_r)
  File "/usr/lib/python3.7/multiprocessing/spawn.py", line 114, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.7/multiprocessing/spawn.py", line 220, in prepare
    set_start_method(data['start_method'], force=True)
  File "/usr/lib/python3.7/multiprocessing/context.py", line 246, in set_start_method
    self._actual_context = self.get_context(method)
  File "/usr/lib/python3.7/multiprocessing/context.py", line 238, in get_context
    return super().get_context(method)
  File "/usr/lib/python3.7/multiprocessing/context.py", line 192, in get_context
    raise ValueError('cannot find context for %r' % method) from None
ValueError: cannot find context for 'loky'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 253, in __call__
    for func, args, kwargs in self.items]
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 253, in <listcomp>
    for func, args, kwargs in self.items]
  File "MNIST_hydra.py", line 22, in main
    for batch in data_loader:
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
    data = self._next_data()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 974, in _next_data
    idx, data = self._get_data()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 941, in _get_data
    success, data = self._try_get_data()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 792, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str))
RuntimeError: DataLoader worker (pid(s) 11427) exited unexpectedly
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 1042, in __call__
    self.retrieve()
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/parallel.py", line 921, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/opt/.virtualenvs/l2r/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
RuntimeError: DataLoader worker (pid(s) 11427) exited unexpectedly
```

**Comment by albanD** [comment]
This does look like different multiprocessing backends that are not compatible and conflict.
I can leave this open in case other people have the same issue: using `fork` seems to be a good workaround for now though.

And if someone is looking into this, we would be happy to update pytorch to make it work nicely with 'loky' but no core dev is looking into this right now.

**Comment by tomMoral** [comment]
Hello,

Just came across this issue. It seems that the error is because the `Dataloader` does not use the proper context when creating the process. It uses a default `multiprocessing.Process` which is not compatible with the loky backend. The following script fixes the issues:


```python
import os, sys
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from joblib import Parallel, delayed
from joblib.externals.loky.backend.context import get_context


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

def main(num_workers, lr):
    X = torch.randn(20000, 28, 28)
    y = torch.randint(high=10, size=(20000,))
    data_set = torch.utils.data.TensorDataset(X, y)
    data_loader = DataLoader(data_set, batch_size=64, num_workers=num_workers,
                             multiprocessing_context=get_context('loky'))
    model = Model()
    optim = torch.optim.Adam(model.parameters(), lr=lr)

    for batch in data_loader:
        x, y = batch
        y_hat = model(x)
        loss = F.cross_entropy(y_hat, y)
    print(f'completed {lr}')

if __name__ == '__main__':
    num_workers = int(sys.argv[1])
    lr_array = [1e-4, 1e-3, 1e-2]
    Parallel(n_jobs=3, backend='loky')(delayed(main)(num_workers, lr) for lr in lr_array)

```

To fix this directly in pytorch, one would need to fix the way the mutliprocessing contexts are handled in `torch.multiprocessing` by making sure we take the object from the default context. (`loky` registers a multiprocessing context that is compatible with this).

**Comment by tomMoral** [comment]
The PR fixing this has been reverted so we should probably reopen this issue.

I will try to make a new PR and see how to get it accepted this time :)
